File: src/de/espend/idea/php/annotation/doctrine/reference/DoctrineAnnotationTypeProvider.java
package de.espend.idea.php.annotation.doctrine.reference;

import com.intellij.psi.*;
import com.jetbrains.php.lang.PhpLangUtil;
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression;
import de.espend.idea.php.annotation.extension.parameter.AnnotationPropertyParameter;
import de.espend.idea.php.annotation.extension.PhpAnnotationReferenceProvider;
import de.espend.idea.php.annotation.extension.parameter.PhpAnnotationReferenceProviderParameter;
import de.espend.idea.php.annotation.doctrine.reference.references.DoctrineRepositoryReference;
import org.jetbrains.annotations.Nullable;


public class DoctrineAnnotationTypeProvider implements PhpAnnotationReferenceProvider {

    @Nullable
    @Override
    public PsiReference[] getPropertyReferences(AnnotationPropertyParameter annotationPropertyParameter, PhpAnnotationReferenceProviderParameter referencesByElementParameter) {

        if(annotationPropertyParameter.getType() != AnnotationPropertyParameter.Type.PROPERTY_VALUE) {
            return null;
        }

        String propertyName = annotationPropertyParameter.getPropertyName();
        if(propertyName == null || !propertyName.equals("repositoryClass")) {
            return null;
        }

        String presentableFQN = annotationPropertyParameter.getPhpClass().getPresentableFQN();
        if(!PhpLangUtil.equalsClassNames("Doctrine\\ORM\\Mapping\\Entity", presentableFQN)) {
            return null;
        }

        return new PsiReference[] {
            new DoctrineRepositoryReference((StringLiteralExpression) annotationPropertyParameter.getElement())
        };

    }

}

File: src/main/kotlin/de/espend/idea/php/annotation/doctrine/reference/DoctrineAnnotationTypeProvider.kt
package de.espend.idea.php.annotation.doctrine.reference

import com.intellij.psi.PsiReference
import com.jetbrains.php.lang.PhpLangUtil
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression
import de.espend.idea.php.annotation.doctrine.reference.references.DoctrineRepositoryReference
import de.espend.idea.php.annotation.extension.PhpAnnotationReferenceProvider
import de.espend.idea.php.annotation.extension.parameter.AnnotationPropertyParameter
import de.espend.idea.php.annotation.extension.parameter.PhpAnnotationReferenceProviderParameter

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
class DoctrineAnnotationTypeProvider : PhpAnnotationReferenceProvider {
    override fun getPropertyReferences(
        annotationPropertyParameter: AnnotationPropertyParameter,
        referencesByElementParameter: PhpAnnotationReferenceProviderParameter?
    ): Array<PsiReference>? {
        if (annotationPropertyParameter.type != AnnotationPropertyParameter.Type.PROPERTY_VALUE) {
            return null
        }

        if (annotationPropertyParameter.propertyName != "repositoryClass") {
            return null
        }

        val presentableFQN = annotationPropertyParameter.phpClass.presentableFQN
        if (!PhpLangUtil.equalsClassNames("Doctrine\\ORM\\Mapping\\Entity", presentableFQN)) {
            return null
        }

        return arrayOf(
            DoctrineRepositoryReference(annotationPropertyParameter.element as StringLiteralExpression)
        )
    }
}
